ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -9,51 +9,119 @@ import akka.stream.scaladsl.{Flow, Source}
 import akka.stream.Materializer
 import com.digitalasset.http.EndpointsCompanion._
 import com.digitalasset.http.domain.{GetActiveContractsRequest, JwtPayload}
-import com.digitalasset.http.json.SprayJson.JsonReaderError
 import com.digitalasset.http.json.{DomainJsonDecoder, DomainJsonEncoder, SprayJson}
+import SprayJson.JsonReaderError
+import ContractsFetch.InsertDeleteStep
+import com.digitalasset.http.LedgerClientJwt.Terminates
+import util.ApiValueToLfValueConverter.apiValueToLfValue
+import json.JsonProtocol.LfValueCodec.{apiValueToJsValue => lfValueToJsValue}
+import query.ValuePredicate.LfV
 import com.digitalasset.jwt.domain.Jwt
-import com.digitalasset.ledger.api.v1.transaction.Transaction
 import com.digitalasset.ledger.api.{v1 => api}
-import com.digitalasset.ledger.client.binding.offset.LedgerOffsetOrdering
-import com.digitalasset.ledger.client.services.transactions.TransactionClient
+
 import com.typesafe.scalalogging.LazyLogging
+import scalaz.Liskov.<~<
 import scalaz.syntax.show._
+import scalaz.syntax.std.boolean._
+import scalaz.syntax.tag._
 import scalaz.syntax.traverse._
-import scalaz.{-\/, \/, \/-}
+import scalaz.{-\/, \/, \/-, Show}
 import spray.json.{JsObject, JsString, JsValue}
 
-import scala.collection.immutable.Set
-import scala.concurrent.ExecutionContext
+import scala.collection.SeqLike
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success}
 
 object WebSocketService {
-  val heartBeat: String = JsObject("heartbeat" -> JsString("ping")).toString
-  val emptyGetActiveContractsRequest = domain.GetActiveContractsRequest(Set.empty, Map.empty)
-  private val numConns = new java.util.concurrent.atomic.AtomicInteger(0)
+  private type CompiledQueries = Map[domain.TemplateId.RequiredPkg, LfV => Boolean]
+  val heartBeat: String = JsObject("heartbeat" -> JsString("ping")).compactPrint
+
+  private implicit final class `\\/ WSS extras`[L, R](private val self: L \/ R) extends AnyVal {
+    def liftErr[M](f: String => M)(implicit L: Show[L]): M \/ R =
+      self leftMap (e => f(e.shows))
+  }
+
+  private implicit final class `Seq WSS extras`[A, Self](private val self: SeqLike[A, Self])
```

Comment:
If you merge master, you can get it from `com.digitalasset.http.util.Collections.SeqOps#partitionMap`. I have borrowed it from here :)

Reply:
Used factoring in c3e5537e2aecde8437425b615b2f58a501c67967 ; thanks.